### ISSUE TITLE
Consider adding a jbang-catalog

### ISSUE THREAD

**Issue by aalmiray** [moditect/layrry#28]
this would allow running layrry in combination with jbang as

    jbang layrry@moditect --layer-config config.yml

Requires `layrry-cli` being published to a remote Maven repository (Maven Central most likely)

**Comment by gunnarmorling** [comment]
Nice, didn't know of that. Sounds definitely useful, seems it'd allow to get started with Layrry much simpler. // CC @maxandersen.

**Comment by maxandersen** [comment]
Yes - it will even get Java for you.

**Comment by gunnarmorling** [comment]
@aalmiray I think that's even more interesting than the SDKMAN one.
@maxandersen, ah, that's nice. Can I control the version somehow, so that layrry-cli also would come with JDK 15 for instance?

**Comment by maxandersen** [comment]
Yes. Not sure we got that supported in catalog directly. Maybe we should ? /cc @quintesse

But you can just write a wrapper jbang script that looks like:

'''
//JAVA 15+
//DEPS <Larry gav>

class Main {

    ...main calling relevant Larry main class...
}
'''

**Comment by maxandersen** [comment]
Note - I haven't checked if adoptopenjdk has 15 exposed yet but once it has it should just work.

**Comment by gunnarmorling** [comment]
Looks like it's time to get a first Layrry release into Maven Central :)

**Comment by aalmiray** [comment]
@gunnarmorling do you want to go with the standard route and publish to Maven Central directly _or_ create and org at bintray.com, publish to JCenter and sync to Maven Central?

We'd also need a tiny pom for `layrry-cli` with just GAV coordinates and the minimum POM elements accepted by MC (such as name, url, license, scm). No need to publish `layrry-launcher` as it currently stands.

**Comment by gunnarmorling** [comment]
Re publication, whatever is simpler; so far I've always published to Maven Central (Sonatype) directly. I think I still need to request the org.moditect.layrry group id (I did it for org.moditect, but I think that's not enough?). Would you be interested in getting deploy permissions? If so, let me know of your Sonatype user name and I'll request that, too.

**Comment by aalmiray** [comment]
AFAICT if you requested access for `org.moditect` then you do not need to request access for `org.moditect.layrry`. I did the same for `org.kordamp` and I can now publishing any subpackage under it without additional access requests.

My sonatype username is **aalmiray**.

**Comment by quintesse** [comment]
Good idea @maxandersen , I've opened [an issue](https://github.com/jbangdev/jbang/issues/311) for it.

Btw, Java 15 is not yet available at AdoptOpenJDK.

**Comment by gunnarmorling** [comment]
> AFAICT if you requested access for org.moditect then you do not need to request access for org.moditect.layrry. I did the same for org.kordamp and I can now publishing any subpackage under it without additional access requests.

Ah, nice, didn't know that. I should have requested org.* long time ago ;)

> My sonatype username is aalmiray.

Thanks, I'll get you added then.

**Comment by maxandersen** [comment]
if you could get the build to not fail on clean checkout then could have used jitpack.

but it fails: https://jitpack.io/com/github/moditect/layrry/-f369cc80ac-1/build.log

so right now i can't even make a quick pr with jbang-catalog...:)

**Comment by gunnarmorling** [comment]
> invalid flag: --release

Ah, let's see, this requires some Jitpack config. One sec.

**Comment by aalmiray** [comment]
The build currently requires JDK14. Perhaps adding an enforcer rule for this constraint would be in order.

**Comment by gunnarmorling** [comment]
Ok, Jitpack should build with JDK 14 now. Could you try again?

**Comment by maxandersen** [comment]
still fail https://jitpack.io/com/github/moditect/layrry/4ad4797e1d/build.log - error about missing headers.

**Comment by aalmiray** [comment]
Looks like `install-jdk.sh` must be skipped from license header checks as it's script provided by a 3rd party (@sormuras)

**Comment by sormuras** [comment]
Huh? Should work ootb... gimme a second.

**Comment by sormuras** [comment]
Ah -  I see. Yes. What Andres says:

> Looks like `install-jdk.sh` must be skipped from license header checks...

My script, my header. :P

**Comment by gunnarmorling** [comment]
Gasp. install-jdk.sh is excluded from the header check now.

**Comment by sormuras** [comment]
You may also delete the script after sourcing it into the current shell session:

https://github.com/moditect/layrry/blob/4ad4797e1dd18fda4455cb4e44baf082d71aa973/jitpack.yml#L19

...before building the project.

**Comment by sormuras** [comment]
Looks better now: https://jitpack.io/com/github/moditect/layrry/master-SNAPSHOT/build.log

**Comment by maxandersen** [comment]
okey!

so this is now possible: `jbang layrry@maxandersen` 

catalog at: https://github.com/maxandersen/jbang-catalog/blob/master/jbang-catalog.json#L15
wrapper script at: https://github.com/maxandersen/jbang-catalog/blob/master/layrry.java

if you create a https://github.com/moditect/jbang-catalog with a similar catalog+wrapper you can do:

`jbang layrry@moditect`

**Comment by aalmiray** [comment]
Cool! We can start with this approach (snapshot release via jitpack) then update the script with a fixed version once a release is posted to Maven Central 😏

**Comment by gunnarmorling** [comment]
Yeah! So I'm trying, but I'm getting this, @maxandersen:

```
jbang layrry@maxandersen --verbose
[jbang] https://github.com/maxandersen/jbang-catalog/blob/master/-d is not from a trusted source thus not running it automatically.

If you trust the url to be safe to run you can do one of the following:
0) Trust once: Add no trust, just run this time
1) Trust this url in future:
    jbang trust add https://github.com/maxandersen/jbang-catalog/


Any other response will result in exit.

[jbang] Type in your choice (0 or 1) and hit enter. Times out after 10 seconds.
0
[jbang] Could not download https://raw.githubusercontent.com/maxandersen/jbang-catalog/master/-d
[jbang] Run with --verbose for more details
```

**Comment by maxandersen** [comment]
hmm....try again (missing a push)

**Comment by gunnarmorling** [comment]
Yeah, that seems to work! Will do some more testing tomorrow. That's really awesome 🎉

**Comment by gunnarmorling** [comment]
Actually, how do I pass JVM parameter (--enable-preview) with jbang?

**Comment by maxandersen** [comment]
@gunnarmorling see https://github.com/jbangdev/jbang/blob/master/examples/records.java#L3

**Comment by aalmiray** [comment]
![kronk](https://media1.tenor.com/images/5fa5653d39930b0f7d7ae2851f55fd5b/tenor.gif)

**Comment by gunnarmorling** [comment]
@maxandersen and what about runtime options? I got `//JAVA_OPTIONS --enable-preview --release 14`, but seems to make no difference?

@aalmiray ;)

**Comment by gunnarmorling** [comment]
That's what I get right now:

```
[jbang] Dependencies resolved
[jbang] Building jar...
Error: LinkageError occurred while loading main class layrry
	java.lang.UnsupportedClassVersionError: Preview features are not enabled for layrry (class file version 58.65535). Try running with '--enable-preview'
```

So IIUC, it did get compiled, but running fails.

**Comment by gunnarmorling** [comment]
Raaah, so got past that, but a classloader issue:

```
23:49:55.540 [main] INFO  com.example.layrry.links.core.internal.LayrryLinksVerticle - Adding plug-in: PluginDescriptor [name=plugins-layrry-links-membership-1.0.0, moduleLayer=com.example.layrry.links.membership]
Exception in thread "main" java.lang.RuntimeException: Couldn't run module main class
	at org.moditect.layrry.internal.LayersImpl.run(LayersImpl.java:130)
	at org.moditect.layrry.Layrry.main(Layrry.java:53)
	at layrry.main(layrry.java:13)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.moditect.layrry.internal.LayersImpl.run(LayersImpl.java:127)
	... 2 more
Caused by: java.util.ServiceConfigurationError: io.vertx.core.spi.VertxFactory: io.vertx.core.impl.VertxFactoryImpl not a subtype
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:591)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1238)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1266)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1301)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1386)
	at vertx.core@3.9.0/io.vertx.core.ServiceHelper.loadFactories(ServiceHelper.java:55)
	at vertx.core@3.9.0/io.vertx.core.ServiceHelper.loadFactories(ServiceHelper.java:42)
	at vertx.core@3.9.0/io.vertx.core.ServiceHelper.loadFactoryOrNull(ServiceHelper.java:32)
	at vertx.core@3.9.0/io.vertx.core.ServiceHelper.loadFactory(ServiceHelper.java:24)
	at vertx.core@3.9.0/io.vertx.core.Vertx.<clinit>(Vertx.java:598)
	at com.example.layrry.links.core@1.0.0/com.example.layrry.links.core.LayrryLinks.main(LayrryLinks.java:28)
```

I'm giving up for today :)

**Comment by sormuras** [comment]
_"Almost there..."_

**Comment by gunnarmorling** [comment]
Famous last words ;)

>

**Comment by maxandersen** [comment]
hmm - could be layrry expect to be launched in a custom java module specific way ? 

you can see how jbang launches by running with `--verbose`.

until now i only had to deal with modules when working with javafx - maybe some usecase I missed ... in any case we can fix that; just need to grok the issue ;)

**Comment by gunnarmorling** [comment]
@maxandersen, is there any description of what jbang does in terms of class-loading? I'm quite puzzled by that exception above, as really these two classes (service interface and impl) should be loaded via Layrry into the same JMPS layer, using one classloader. This is all controlled by the Layrry launcher main, not quite clear to me yet how jbang could interfere with that.

**Comment by maxandersen** [comment]
jbang does nothing in terms of class-loading (except for its SPI integration which is only engaged using Quarkus at the moment)

jbang just collect the dependency tree as specified via maven and puts it into a call to java.

See below output of `jbang --verbose` (after its already been built)

```
❯ jbang --verbose layrry@maxandersen
[jbang] System Java version detected as 11
[jbang] System Java version 11 incompatible, using Jbang managed version 14
[jbang] run: /Users/max/.jbang/cache/jdks/14/bin/java -classpath /Users/max/.jbang/cache/jars/layrry.java.b09ce52a95f9047c92130aee720742b7dc4e5ec3f8d29cb408bbf6b5dc293dfa.jar:/Users/max/.m2/repository/com/github/moditect/layrry/master-a72cd47cb1-1/layrry-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/greeter-core/master-a72cd47cb1-1/greeter-core-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/greeter-date-util/master-a72cd47cb1-1/greeter-date-util-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/greeter-de/master-a72cd47cb1-1/greeter-de-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/greeter-en/master-a72cd47cb1-1/greeter-en-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/greeter-fr/master-a72cd47cb1-1/greeter-fr-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/it-app/master-a72cd47cb1-1/it-app-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/it-bar/master-a72cd47cb1-1/it-bar-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/org/apache/logging/log4j/log4j-api/2.13.1/log4j-api-2.13.1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/it-foo/master-a72cd47cb1-1/it-foo-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/it-greeter/master-a72cd47cb1-1/it-greeter-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/it-logconfig/master-a72cd47cb1-1/it-logconfig-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-config/master-a72cd47cb1-1/layrry-config-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-config-toml/master-a72cd47cb1-1/layrry-config-toml-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/jezza/toml/1.2/toml-1.2.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-config-yaml/master-a72cd47cb1-1/layrry-config-yaml-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/org/yaml/snakeyaml/1.26/snakeyaml-1.26.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-integrationtest-runner/master-a72cd47cb1-1/layrry-integrationtest-runner-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-launcher/master-a72cd47cb1-1/layrry-launcher-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/org/slf4j/slf4j-simple/1.7.29/slf4j-simple-1.7.29.jar:/Users/max/.m2/repository/org/slf4j/slf4j-api/1.7.29/slf4j-api-1.7.29.jar:/Users/max/.m2/repository/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-impl-maven/3.1.4/shrinkwrap-resolver-impl-maven-3.1.4.jar:/Users/max/.m2/repository/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-spi-maven/3.1.4/shrinkwrap-resolver-spi-maven-3.1.4.jar:/Users/max/.m2/repository/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-spi/3.1.4/shrinkwrap-resolver-spi-3.1.4.jar:/Users/max/.m2/repository/org/apache/maven/maven-resolver-provider/3.6.3/maven-resolver-provider-3.6.3.jar:/Users/max/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar:/Users/max/.m2/repository/org/apache/maven/maven-model/3.6.3/maven-model-3.6.3.jar:/Users/max/.m2/repository/org/apache/maven/maven-model-builder/3.6.3/maven-model-builder-3.6.3.jar:/Users/max/.m2/repository/org/apache/maven/maven-artifact/3.6.3/maven-artifact-3.6.3.jar:/Users/max/.m2/repository/org/apache/commons/commons-lang3/3.8.1/commons-lang3-3.8.1.jar:/Users/max/.m2/repository/org/apache/maven/maven-builder-support/3.6.3/maven-builder-support-3.6.3.jar:/Users/max/.m2/repository/org/eclipse/sisu/org.eclipse.sisu.inject/0.3.4/org.eclipse.sisu.inject-0.3.4.jar:/Users/max/.m2/repository/org/apache/maven/maven-repository-metadata/3.6.3/maven-repository-metadata-3.6.3.jar:/Users/max/.m2/repository/org/apache/maven/maven-settings/3.6.3/maven-settings-3.6.3.jar:/Users/max/.m2/repository/org/apache/maven/maven-settings-builder/3.6.3/maven-settings-builder-3.6.3.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-api/1.4.1/maven-resolver-api-1.4.1.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-impl/1.4.1/maven-resolver-impl-1.4.1.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-spi/1.4.1/maven-resolver-spi-1.4.1.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-util/1.4.1/maven-resolver-util-1.4.1.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-connector-basic/1.4.1/maven-resolver-connector-basic-1.4.1.jar:/Users/max/.m2/repository/org/apache/maven/resolver/maven-resolver-transport-wagon/1.4.1/maven-resolver-transport-wagon-1.4.1.jar:/Users/max/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.25/plexus-interpolation-1.25.jar:/Users/max/.m2/repository/org/codehaus/plexus/plexus-utils/3.2.1/plexus-utils-3.2.1.jar:/Users/max/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.4/plexus-sec-dispatcher-1.4.jar:/Users/max/.m2/repository/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar:/Users/max/.m2/repository/org/apache/maven/wagon/wagon-provider-api/3.3.4/wagon-provider-api-3.3.4.jar:/Users/max/.m2/repository/org/apache/maven/wagon/wagon-file/3.3.4/wagon-file-3.3.4.jar:/Users/max/.m2/repository/org/apache/maven/wagon/wagon-http-lightweight/2.12/wagon-http-lightweight-2.12.jar:/Users/max/.m2/repository/org/apache/maven/wagon/wagon-http-shared/2.12/wagon-http-shared-2.12.jar:/Users/max/.m2/repository/org/jsoup/jsoup/1.7.2/jsoup-1.7.2.jar:/Users/max/.m2/repository/commons-io/commons-io/2.5/commons-io-2.5.jar:/Users/max/.m2/repository/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-api-maven/3.1.4/shrinkwrap-resolver-api-maven-3.1.4.jar:/Users/max/.m2/repository/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-api/3.1.4/shrinkwrap-resolver-api-3.1.4.jar:/Users/max/.m2/repository/com/beust/jcommander/1.78/jcommander-1.78.jar:/Users/max/.m2/repository/io/methvin/directory-watcher/0.9.9/directory-watcher-0.9.9.jar:/Users/max/.m2/repository/net/java/dev/jna/jna/5.5.0/jna-5.5.0.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-links-core/master-a72cd47cb1-1/layrry-links-core-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/io/vertx/vertx-core/3.9.0/vertx-core-3.9.0.jar:/Users/max/.m2/repository/io/netty/netty-common/4.1.48.Final/netty-common-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-buffer/4.1.48.Final/netty-buffer-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-transport/4.1.48.Final/netty-transport-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-handler/4.1.48.Final/netty-handler-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-codec/4.1.48.Final/netty-codec-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-handler-proxy/4.1.48.Final/netty-handler-proxy-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-codec-socks/4.1.48.Final/netty-codec-socks-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-codec-http/4.1.48.Final/netty-codec-http-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-codec-http2/4.1.48.Final/netty-codec-http2-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-resolver/4.1.48.Final/netty-resolver-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-resolver-dns/4.1.48.Final/netty-resolver-dns-4.1.48.Final.jar:/Users/max/.m2/repository/io/netty/netty-codec-dns/4.1.48.Final/netty-codec-dns-4.1.48.Final.jar:/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.10.2/jackson-core-2.10.2.jar:/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.10.2/jackson-databind-2.10.2.jar:/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.10.2/jackson-annotations-2.10.2.jar:/Users/max/.m2/repository/io/vertx/vertx-web/3.9.0/vertx-web-3.9.0.jar:/Users/max/.m2/repository/io/vertx/vertx-web-common/3.9.0/vertx-web-common-3.9.0.jar:/Users/max/.m2/repository/io/vertx/vertx-auth-common/3.9.0/vertx-auth-common-3.9.0.jar:/Users/max/.m2/repository/io/vertx/vertx-bridge-common/3.9.0/vertx-bridge-common-3.9.0.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-links-logconfig/master-a72cd47cb1-1/layrry-links-logconfig-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-links-membership/master-a72cd47cb1-1/layrry-links-membership-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-links-runner/master-a72cd47cb1-1/layrry-links-runner-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-links-tournament/master-a72cd47cb1-1/layrry-links-tournament-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-platform/master-a72cd47cb1-1/layrry-platform-master-a72cd47cb1-1.jar:/Users/max/.m2/repository/com/github/moditect/layrry/layrry-plugin-example-runner/master-a72cd47cb1-1/layrry-plugin-example-runner-master-a72cd47cb1-1.jar layrry
Exception in thread "main" com.beust.jcommander.ParameterException: The following option is required: [--layers-config]
	at com.beust.jcommander.JCommander.validateOptions(JCommander.java:388)
	at com.beust.jcommander.JCommander.parse(JCommander.java:357)
	at com.beust.jcommander.JCommander.parse(JCommander.java:335)
	at org.moditect.layrry.Layrry.main(Layrry.java:42)
	at layrry.main(layrry.java:11)
```

**Comment by sormuras** [comment]
Uff. That looooong class path...

FYI, jbang + bach just works: 

```
G:\project>copy con module-info.java
module project {}
^Z
        1 Datei(en) kopiert.

G:\project>D:\nosetup\jbang-0.45.0\bin\jbang de.sormuras.bach:de.sormuras.bach:11.8
[jbang] Resolving dependencies...
[jbang]     Resolving de.sormuras.bach:de.sormuras.bach:11.8...Done
[jbang] Dependencies resolved
Build of project project 1-ea started by Bach.java 11.8

Build main space
Compile module project
Create archive project@1-ea.jar
Create archive project@1-ea-sources.jar
Generate API documentation for module project
Create archive project@1-ea-api.jar

Tool Call Overview contains 5 distinct calls
  Duration       Name Arguments
     0.25s      javac --module project --module-version 1-ea --module-source-path project=. -encoding UTF-8 -paramete...
     0.05s        jar --create --file .bach\workspace\sources\project@1-ea-sources.jar --no-manifest -C . .
     0.05s        jar --create --file .bach\workspace\modules\project@1-ea.jar -C .bach\workspace\classes\15\project .
      0.2s    javadoc --module project --module-source-path project=. -d .bach\workspace\documentation\api -encoding ...
    0.141s        jar --create --file .bach\workspace\documentation\project@1-ea-api.jar --no-manifest -C .bach\works...
  -------- ----------
    0.691s    5 calls

Directory modules (file:///G:/project/.bach/workspace/modules/) contains 1 file
- project@1-ea.jar with 587 bytes

Logbook written to file:///G:/project/.bach/workspace/logbook.md
Build of project project 1-ea took 698 ms
```

**Comment by gunnarmorling** [comment]
@maxandersen, interesting, where's this classpath coming from? Is it examining the POM in the current directory somehow? This really should only contain the Layrry launcher JAR (with its main class), whereas everything else (e.g. the greeter and layrry-links example modules) will be loaded by Layrry via module layers itself. So this explains the exception I got, question is how that classpath is created.

**Comment by sormuras** [comment]
I bet, that's due to Layrry being pulled from JitPack -- which "install"s all available Maven artifacts:

```text
Build artifacts:
com.github.moditect.layrry:layrry-config-toml:master-a72cd47cb1-1
com.github.moditect.layrry:greeter-fr:master-a72cd47cb1-1
com.github.moditect.layrry:it-greeter:master-a72cd47cb1-1
com.github.moditect.layrry:it-foo:master-a72cd47cb1-1
com.github.moditect.layrry:it-app:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-plugin-example-runner:master-a72cd47cb1-1
com.github.moditect.layrry:greeter-de:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-example:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-platform:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-config-yaml:master-a72cd47cb1-1
com.github.moditect.layrry:greeter-date-util:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-runner:master-a72cd47cb1-1
com.github.moditect.layrry:greeter-core:master-a72cd47cb1-1
com.github.moditect.layrry:it-logconfig:master-a72cd47cb1-1
com.github.moditect.layrry:greeter-en:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-plugin-example:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-core:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-launcher:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-integrationtest:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-config:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-integrationtest-runner:master-a72cd47cb1-1
com.github.moditect.layrry:it-bar:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-tournament:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-aggregator:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-logconfig:master-a72cd47cb1-1
com.github.moditect.layrry:layrry-links-membership:master-a72cd47cb1-1
```

Copied from https://jitpack.io/com/github/moditect/layrry/master-a72cd47cb1-1/build.log

**Comment by gunnarmorling** [comment]
Hum, hum. Installing should be fine, still not sure why/how everything ends up on jbang's classpath? Essentially, what needs to run is the equivalent to

    java --enable-preview -jar layrry-cli-1.0-SNAPSHOT.jar

**Comment by maxandersen** [comment]
i've updated the catalog to just use com.github.moditect.layrry:layrry-launcher:master-SNAPSHOT

let me know if that works better.

if not, we might have the first usecase of finding a way to exclude dependencies...

**Comment by gunnarmorling** [comment]
```
jbang layrry@moditect --layers-config layrry-links-runner/src/test/resources/layers.yml
...
10:04:03.165 [vert.x-eventloop-thread-0] INFO  c.e.l.l.c.i.LayrryLinksVerticle - Adding `router` for path: /routes
10:04:03.176 [vert.x-eventloop-thread-0] INFO  c.e.l.l.c.i.LayrryLinksVerticle - Adding router for path: /members
10:04:03.299 [vert.x-eventloop-thread-0] INFO  c.e.l.l.c.i.LayrryLinksVerticle - Server ready! Browse to http://localhost:8080/routes
```

Woot  🎉 !

**Comment by gunnarmorling** [comment]
So that's really, really sweet; solves distribution of the Layrry launcher, so people can get started with layered apps within minutes (Layrry resolves all dependencies of the layered application itself).

**Comment by gunnarmorling** [comment]
The catalog is here: https://github.com/moditect/jbang-catalog.

@aalmiray, I think this resolves this issue? We can tackle using the released version separately IMO. For now I think it's even better to work off of master via jitpack, allowing fixes to be published and consumed quickly. WDYT?

**Comment by sormuras** [comment]
> Layrry resolves all dependencies of the layered application itself

This is the interesting part - where Bach and Layrry meet. Let's how both projects may join forces here (in a new issue).

**Comment by maxandersen** [comment]
excellent - happy it worked :) 

btw. the `jbang <catalog>` syntax lets you do all kind of stuff like referring to sub dirs, other names and branches - see it at https://github.com/jbangdev/jbang#implicit-alias-catalogs

or you could just add a `layrry-master` alias in the catalog to allow referring to both a stable or not version.

all upto what works best for the way you want to distribute it.

**Comment by gunnarmorling** [comment]
> This is the interesting part - where Bach and Layrry meet. Let's how both projects may join forces here (in a new issue).

Yeah, let's discuss!

> or you could just add a layrry-master alias in the catalog to allow referring to both a stable or not version.

Cool, would have been my next question. I'll check it out. For now I think using master is ideal, but we may want to move to a stable version once things have settled a bit. Right now I love the fact that I don't need to deal with releases to Maven Central that way :)

**Comment by aalmiray** [comment]
Agreed. This makes it easier to test out Layrry, only requirement being `jbang` installed and perhaps a JRE if it's not pulled by jbang. This of course assumes network access. 

On a separate issue I'd like to discuss offline mode (local repo which may or may not be Maven Local).

**Comment by gunnarmorling** [comment]
> On a separate issue I'd like to discuss offline mode (local repo which may or may not be Maven Local).

+1 Good idea!

**Comment by gunnarmorling** [comment]
🎈 Closing this one then; thanks everyone, really cool stuff!

**Comment by maxandersen** [comment]
Small correction : it does require a jdk as jbang needs to be able to call javac to compile the wrapper .Java script.

**Comment by aalmiray** [comment]
We're only missing documenting in the README before we can close this issue 😏
